src/detect.c

```diff
@@ -144,7 +144,8 @@ static void DetectRun(ThreadVars *th_v,
     PACKET_PROFILING_DETECT_END(p, PROF_DETECT_RULES);
 
     /* run tx/state inspection. Don't call for ICMP error msgs. */
-    if (pflow && pflow->alstate && likely(pflow->proto == p->proto)) {
+    if (pflow && pflow->alstate && likely(pflow->proto == p->proto) &&
+            (p->flags & PKT_STREAM_EST)) {
```

Comment:
one thing I would like to know here is if TCP fast open, data on syn, IPS mode (so app-layer is called immediately) also sets this flag

Reply:
should a similar check be done before the tx logging is called?

Reply:
> should a similar check be done before the tx logging is called?

There are likely other places where this optimization can happen

Reply:
also the flag will only be set for TCP I think

Reply:
> one thing I would like to know here is if TCP fast open, data on syn, IPS mode (so app-layer is called immediately) also sets this flag

I do not think it does :
Code should get into `StreamTcpPacketStateNone`, case `if (p->tcph->th_flags & TH_SYN)` and `TCP_HAS_TFO` so not going to set `PKT_STREAM_EST` because not in same code path as `StreamTcpPacketStateNone`